refactor: propagate version formatting errors

vladimyr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
<!--- Describe your changes in detail -->
Propagate errors originating from `VersionFormatter#format_version` to the upper level.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Continuation of #2499
(It got closed before I managed to leave my review.)

#### Screenshots (if appropriate):

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

---

/cc @offbyone

offbyone commented on src/modules/java.rs:
```diff
@@ -74,19 +71,24 @@ fn get_java_version(context: &Context, config: &JavaConfig) -> Option<String> {
+
+    match formatted {
+        Ok(formatted) => Some(formatted),
+        Err(error) => {
```
What conditions would lead to this warning? I would rather just silently format the raw version than log a warning when this happens.

vladimyr replied:
> I would rather just silently format the raw version than log a warning when this happens.

This is _silent_ unless you explicitly opt-in by setting `STARSHIP_LOG` environment variable to the `warn` level.

> What conditions would lead to this warning?

The simplest case that would cause an error is a malformed version format like:
```toml
version = "v${raw" # missing right curly
```

This results in (with logging enabled):
```
[WARN] - (starship::modules::java): Error formating `java` version:
 --> 1:3
  |
1 | v${raw
  |   ^---
  |
  = expected variable_name
```

offbyone reviewed (approved):
I'm mostly curious about the error handling rationale; I personally would prefer that version formatting be fairly silent, if possible, but that might be a "just me" philosophy. Otherwise, it looks like you reordered parameters in a few places, but ...  ¯\_(ツ)_/¯ different strokes for different folks :)

vladimyr commented:
> I'm mostly curious about the error handling rationale; I personally would prefer that version formatting be fairly silent, if possible, but that might be a "just me" philosophy.

As I explained above https://github.com/starship/starship/pull/2566#discussion_r610245095 this is still silent. Error handling is moved to the upper level because version formatter doesn't know who called it thus can't print contextually rich warnings.

> Otherwise, it looks like you reordered parameters in a few places, but ... ¯_(ツ)_/¯ different strokes for different folks :)

This speaks for itself ie I find the original version so good I didn't feel a need to change things around :wink:
I did reordering to achieve the following form: `format_version(version /*[1]*/, format /*[2]*/)` where version is followed by format so it reads like: _version formatter formats **version [1]** using provided **format [2]**_.

vladimyr commented:
> Otherwise, it looks like you reordered parameters in a few places

Actually, there is more than just that. I did 3 additional things there:
1. ensured that every `format_{module}_version` function has following signature `format_{module}_version(version, version_string)` (no context, config, just format string)
2. simplified version parsing in a few of those
3. renamed a bunch of java tests from `test_parsing_*` to `test_format_*` in order to follow already established convention

vladimyr commented on src/formatter/version.rs:
```diff
@@ -17,81 +19,85 @@ impl<'a> VersionFormatter<'a> {
-                .join(""),
-            Err(_) => version.to_string(),
-        }
+                .join("")
```
Done, thanks for the tip! :+1:
